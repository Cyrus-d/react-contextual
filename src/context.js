import React from 'react'
import { createContext } from 'react-broadcast'

const providers = new Map()
const Context = createContext()

export function createNamedContext(name, initialState) {
    const context = createContext(initialState)
    providers.set(name, context)
    return context
}

export function getNamedContext(name) {
    return providers.get(name)
}

export function removeNamedContext(name) {
    providers.delete(name)
}

export function resolveContext(context, props) {
    let result
    if (typeof context === 'function') {
        // Test against component-symbol first, then assume a user function
        result = getNamedContext(context) || resolveContext(context(props))
    } else if (typeof context === 'string') {
        result = getNamedContext(context)
    }
    return result || context ||Â Context
}

export default Context