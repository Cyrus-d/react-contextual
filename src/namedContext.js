import React from 'react'
import {
  resolveContext,
  createNamedContext,
  removeNamedContext,
} from './context'

export default function namedContext(contextName, initialState) {
  return Wrapped =>
    class NamedContext extends React.PureComponent {
      constructor(props) {
        super()
        this.name =
          typeof contextName === 'function' ? contextName(props) : contextName
        this.state = { context: createNamedContext(this.name, initialState) }
      }
      componentWillUnmount() {
        removeNamedContext(this.name)
      }
      render() {
        return <Wrapped {...this.props} context={this.state.context} />
      }
    }
}
