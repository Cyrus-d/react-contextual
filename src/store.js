import React from 'react'
import PropTypes from 'prop-types'
import uuid from 'tiny-uuid'
import ProviderContext, { createNamedContext, removeNamedContext } from './context'

export function createStore(data, id = uuid()) {
    const { initialState, actions, ...props } = data
    const subscriptions = new Set()
    const result = {
        subscriptions,
        initialState,
        state: initialState,
        actions,
        props,
        id,
        context: createNamedContext(id),
        destroy: () => {
            removeNamedContext(id)
            subscriptions.clear()
        },
        subscribe: callback => {
            subscriptions.add(callback)
            return () => subscriptions.delete(callback)
        },
        getState: () => result.state
    }
    return result
}

export class RenderPure extends React.PureComponent {
    render() {
        return this.props.children
    }
}

export class Provider extends React.PureComponent {
    static propTypes = {
        id: PropTypes.string,
        initialState: PropTypes.object,
        actions: PropTypes.object,
        store: PropTypes.object,
    }

    update = done => {
        // Update store
        this.store.state = this.state
        // Call subscribers
        this.store.subscriptions.forEach(value => value(this.store.state))
        // Resolve
        done()
    }

    constructor(props) {
        super()
        const { store, children, id, actions, initialState, ...rest } = props
        this.store = store || createStore({ ...rest, initialState, actions }, id)
        // When no store is given, create context by id or refer to the default context
        if (!store) this.store.context = id ? createNamedContext(id) : ProviderContext
        // Bind actions to setState
        if (this.store.actions) {
            const actions = this.store.actions
            this.store.actions = Object.keys(actions).reduce(
                (acc, name) => ({
                    ...acc,
                    [name]: (...args) =>
                        new Promise(res => {
                            let result = actions[name](...args)
                            if (typeof result === 'function')
                                Promise.resolve(result(this.state)).then(state =>
                                    this.setState(state, () => this.update(res)),
                                )
                            else this.setState(result, () => this.update(res))
                        }),
                }),
                {},
            )
        }
        this.state = this.store.initialState
    }
    componentWillUnmount() {
        if (this.props.id) removeNamedContext(this.props.id)
    }
    render() {
        const { actions, props } = this.store
        const value = { ...props, ...this.state, ...(actions ? { actions } : {}) }
        return <this.store.context.Provider value={value} children={<RenderPure children={this.props.children} />} />
    }
}
