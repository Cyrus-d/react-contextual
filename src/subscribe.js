import React from 'react'
import PropTypes from 'prop-types'
import ProviderContext, { getNamedContext, resolveContext } from './context'

const toArray = obj => Array.isArray(obj) ? obj : [obj]

export function subscribe(...args) {
    return Wrapped =>
        function SubscribeWrap(props) {
            // Filter undefined args (can happen if Subscribe injects them)
            args = args.filter(a => a)
            let contextRefs = ProviderContext,
                mapContextToProps = props => props
            if (args.length === 1) {
                // Check if the argument is a valid context first, if not, assume mapContextToProps
                if (resolveContext(args[0], props, null)) contextRefs = args[0]
                else mapContextToProps = args[0]
            } else if (args.length === 2) {
                // subscribe(Context, mapContextToProps)
                contextRefs = args[0]
                mapContextToProps = args[1]
            }
            if (typeof mapContextToProps !== 'function') {
                // 'theme' or ['theme', 'user', 'language']
                const values = mapContextToProps
                mapContextToProps = (...args) =>
                    toArray(values).reduce((acc, key, index) => ({ ...acc, [key]: args[index] }), {})
            }
            contextRefs = toArray(contextRefs).map(context => resolveContext(context, props))
            let values = []
            return [...contextRefs, Wrapped].reduceRight((accumulator, Context, i) => (
                <Context.Consumer>
                    {value => {
                        values[i] = value
                        if (accumulator === Wrapped) {
                            let context = mapContextToProps(...values, props)
                            context = typeof context === 'object' ? context : { context }
                            values = []
                            return <Wrapped {...props} {...context} />
                        } else return accumulator
                    }}
                </Context.Consumer>
            ))
        }
}

export class Subscribe extends React.PureComponent {
    static propTypes = {
        to: PropTypes.oneOfType([
            PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.object, PropTypes.string, PropTypes.func])),
            PropTypes.object,
            PropTypes.string,
            PropTypes.func,
        ]),
        select: PropTypes.oneOfType([PropTypes.func, PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
        children: PropTypes.func.isRequired,
    }
    static defaultProps = { to: ProviderContext, select: props => props }
    render() {
        const { to, select, children, ...rest } = this.props
        return React.createElement(subscribe(to, select)(props => children(props)), rest)
    }
}
