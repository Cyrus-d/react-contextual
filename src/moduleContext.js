import React from 'react'
import { createNamedContext } from './context'

export default function moduleContext(initialState) {
    return Wrapped => {
        let context = undefined
        const ModuleContext = class extends React.PureComponent {
            render() {
                return <Wrapped {...this.props} context={context} />
            }
        }
        context = createNamedContext(ModuleContext, initialState)
        return ModuleContext
    }
}