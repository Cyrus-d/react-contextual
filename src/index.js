import { subscribe, Subscribe } from './subscribe'
import { RenderPure, Provider } from './store'
import Context, {
    namedContext,
    moduleContext,
    transformContext,
    createNamedContext,
    removeNamedContext,
    getNamedContext,
    resolveContext,
} from './context'

export {
    Context,
    subscribe,
    Subscribe,
    RenderPure,
    Provider,
    namedContext,
    moduleContext,
    transformContext,
    createNamedContext,
    removeNamedContext,
    getNamedContext,
}
