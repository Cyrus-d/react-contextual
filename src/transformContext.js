import React from 'react'
import { subscribe } from './subscribe'
import { resolveContext } from './context'

export default function transformContext(context, transform) {
    return Wrapped => {
        return subscribe(context, transform)(class extends React.PureComponent {
            render() {
                const context = resolveContext(context, this.props)
                return <Wrapped {...this.props} context={context} />
            }
        })
    }
}