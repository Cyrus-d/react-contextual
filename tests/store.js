import React from 'react'
import delay from 'delay'
import { createStore, Provider, ProviderContext, Subscribe, subscribe, getNamedContext } from '../src/'

const store = {
    initialState: { count: 0 },
    actions: {
        simple: arg => ({ count: arg }),
        functional: arg => state => ({ count: state.count + arg }),
        async: arg => async state => {
            await delay(1)
            return { count: state.count + arg }
        },
    },
    extra: 100,
}

it('renders properly', async () => {
    const Test = subscribe()(props => props.count)
    await snapshot(
        <Provider initialState={{ count: 0 }}>
            <Test />
        </Provider>,
    )
})

test('mount/unmount', async () => {
    const Test = subscribe('store', props => props)(props => props.count)
    const App = class extends React.PureComponent {
        state = { show: true }
        componentDidMount() {
            expect(getNamedContext('store')).toBeTruthy()
            this.setState({ show: false })
        }
        componentDidUpdate() {
            expect(getNamedContext('store')).toBe(undefined)
        }
        render() {
            return this.state.show ? (
                <Provider {...store}>
                    <Provider id="store" {...store}>
                        <Test />
                    </Provider>
                </Provider>
            ) : null
        }
    }
    await snapshot(<App />)
})

test('no actions', async () => {
    const Test = subscribe()(props => props.count)
    await snapshot(
        <Provider initialState={{ count: 0 }} actions={null}>
            <Test />
        </Provider>,
    )
})

test('no actions, setState', async () => {
    const Test = subscribe()(props => (
        <button onClick={() => props.actions.setState(state => ({ count: state.count + 1 }))}>{props.count}</button>
    ))
    await snapshot(
        <Provider initialState={{ count: 0 }}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('simple action', async () => {
    const Test = subscribe()(props => <button onClick={() => props.actions.simple(1)}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('functional action', async () => {
    const Test = subscribe()(props => <button onClick={() => props.actions.functional(1)}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('async action', async () => {
    const Test = subscribe()(props => <button onClick={() => props.actions.async(1)}>{props.count}</button>)
    await snapshot(
        <Provider {...store}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('await actions & access state', async () => {
    const Test = subscribe()(
        class extends React.PureComponent {
            render() {
                return (
                    <div>
                        {this.props.count}
                        <button
                            onClick={async () => {
                                await this.props.actions.simple(this.props.count + 1)
                                await this.props.actions.functional(this.props.count)
                                await this.props.actions.async(this.props.count)
                            }}
                        />
                    </div>
                )
            }
        },
    )
    await snapshot(
        <Provider {...store}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('external store', async () => {
    const externalStore = createStore(store, 'externalTest')
    const Test = subscribe(externalStore, props => ({ count: props.count }))(props => (
        <button onClick={() => externalStore.actions.async(1)}>{props.count}</button>
    ))
    await snapshot(
        <Provider store={externalStore}>
            <Test />
        </Provider>,
        async () => {
            await externalStore.actions.async(1)
            expect(externalStore.getState().count).toBe(1)
            const remove = externalStore.subscribe(state => expect(state.count).toBe(2))
            await externalStore.actions.async(1)
            remove()
            externalStore.destroy()
            expect(externalStore.subscriptions.size).toBe(0)
            expect(getNamedContext('externalTest')).toBeUndefined()
        },
    )
})

test('external store, setState', async () => {
    const externalStore = createStore({ initialState: { count: 0 } }, 'externalTest2')
    const Test = subscribe(externalStore)(props => (
        <button onClick={() => props.actions.setState(state => ({ count: state.count + 1 }))}>{props.count}</button>
    ))
    await snapshot(
        <Provider store={externalStore}>
            <Test />
        </Provider>,
        tree => tree.find('button').simulate('click'),
    )
})

test('extras', async () => {
    const externalStore = createStore({ initialState: { count: 5 }, extra: 1000 }, 'externalTest3')
    const Test = subscribe([ProviderContext, externalStore], (a, b) => ({ a, b }))(
        props => props.a.count + props.a.extra + props.b.count + props.b.extra,
    )
    await snapshot(
        <Provider {...store}>
            <Provider store={externalStore}>
                <Test />
            </Provider>
        </Provider>,
    )
})
