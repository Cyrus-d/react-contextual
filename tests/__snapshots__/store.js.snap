// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async action 1`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`async action 2`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`await actions & access state 1`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <_class2
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={0}
          >
            <div>
              0
              <button
                onClick={[Function]}
              />
            </div>
          </_class2>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`await actions & access state 2`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 4,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <_class2
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={4}
          >
            <div>
              4
              <button
                onClick={[Function]}
              />
            </div>
          </_class2>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`external store 1`] = `
<Provider
  store={
    Object {
      "actions": Object {
        "async": [Function],
        "functional": [Function],
        "simple": [Function],
      },
      "context": Object {
        "Consumer": [Function],
        "Provider": [Function],
      },
      "destroy": [Function],
      "getState": [Function],
      "id": "externalTest",
      "initialState": Object {
        "count": 0,
      },
      "props": Object {},
      "state": Object {
        "count": 0,
      },
      "subscribe": [Function],
      "subscriptions": Set {},
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`external store 2`] = `
<Provider
  store={
    Object {
      "actions": Object {
        "async": [Function],
        "functional": [Function],
        "simple": [Function],
      },
      "context": Object {
        "Consumer": [Function],
        "Provider": [Function],
      },
      "destroy": [Function],
      "getState": [Function],
      "id": "externalTest",
      "initialState": Object {
        "count": 0,
      },
      "props": Object {},
      "state": Object {
        "count": 2,
      },
      "subscribe": [Function],
      "subscriptions": Set {},
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 2,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            count={2}
          >
            <button
              onClick={[Function]}
            >
              2
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`functional action 1`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`functional action 2`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`mount/unmount 1`] = `<App />`;

exports[`renders properly 1`] = `
<Provider
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            count={0}
          >
            0
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`simple action 1`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 0,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={0}
          >
            <button
              onClick={[Function]}
            >
              0
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;

exports[`simple action 2`] = `
<Provider
  actions={
    Object {
      "async": [Function],
      "functional": [Function],
      "simple": [Function],
    }
  }
  initialState={
    Object {
      "count": 0,
    }
  }
>
  <Provider
    value={
      Object {
        "actions": Object {
          "async": [Function],
          "functional": [Function],
          "simple": [Function],
        },
        "count": 1,
      }
    }
  >
    <RenderPure>
      <Component>
        <Consumer
          quiet={false}
        >
          <Component
            actions={
              Object {
                "async": [Function],
                "functional": [Function],
                "simple": [Function],
              }
            }
            count={1}
          >
            <button
              onClick={[Function]}
            >
              1
            </button>
          </Component>
        </Consumer>
      </Component>
    </RenderPure>
  </Provider>
</Provider>
`;
